File: artifacts/libzahl-1.0/refactor/zptest.c
#include <errno.h>
#include <internals.h>
#include <setjmp.h>
#include <stddef.h>
#include <stdint.h>
#include <stdlib.h>
#include <string.h>
#include <zahl.h>
#include <zptest.c>

typedef uint32_t zahl_char_t;
typedef struct 
{
  int sign;
  size_t used;
  size_t alloced;
  zahl_char_t *chars;
} z_t[1];
enum zprimality
{
  NONPRIME = 0,
  PROBABLY_PRIME,
  PRIME
};
enum zranddev
{
  FAST_RANDOM = 0,
  SECURE_RANDOM
};
enum zranddist
{
  QUASIUNIFORM = 0,
  UNIFORM
};
enum zerror
{
  ZERROR_ERRNO_SET = 0
};
void zsetup(jmp_buf);
void zunsetup(void);
void zfree(z_t);
void zswap(z_t, z_t);
size_t zsave(z_t, void *);
size_t zload(z_t, const void *);
void zset(z_t, z_t);
void zseti(z_t, long long int);
void zsetu(z_t, unsigned long long int);
int zcmp(z_t, z_t);
int zcmpi(z_t, long long int);
int zcmpu(z_t, unsigned long long int);
int zcmpmag(z_t, z_t);
void zadd(z_t, z_t, z_t);
void zsub(z_t, z_t, z_t);
void zmul(z_t, z_t, z_t);
void zmodmul(z_t, z_t, z_t, z_t);
void zdiv(z_t, z_t, z_t);
void zdivmod(z_t, z_t, z_t, z_t);
void zmod(z_t, z_t, z_t);
void zsqr(z_t, z_t);
void zmodsqr(z_t, z_t, z_t);
void zneg(z_t, z_t);
void zabs(z_t, z_t);
void zpow(z_t, z_t, z_t);
void zmodpow(z_t, z_t, z_t, z_t);
void zpowu(z_t, z_t, unsigned long long int);
void zmodpowu(z_t, z_t, unsigned long long int, z_t);
void zadd_unsigned(z_t, z_t, z_t);
void zsub_unsigned(z_t, z_t, z_t);
void zand(z_t, z_t, z_t);
void zor(z_t, z_t, z_t);
void zxor(z_t, z_t, z_t);
void znot(z_t, z_t);
void zlsh(z_t, z_t, size_t);
void zrsh(z_t, z_t, size_t);
void ztrunc(z_t, z_t, size_t);
int zbtest(z_t, size_t);
void zsplit(z_t, z_t, z_t, size_t);
size_t zbits(z_t);
size_t zlsb(z_t);
void zbset(z_t, z_t, size_t, int);
void zgcd(z_t, z_t, z_t);
enum zprimality zptest(z_t, z_t, int);
void zrand(z_t, enum zranddev, enum zranddist, z_t);
char *zstr(z_t, char *);
int zsets(z_t, const char *);
size_t zstr_length(z_t, unsigned long long int);
enum zerror zerror(const char **);
void zperror(const char *);
inline static void zinit(z_t a)
{
  a->alloced = 0;
  a->chars = 0;
}

inline static int zeven(z_t a)
{
  return (!a->sign) || (!(a->chars[0] & 1));
}

inline static int zodd(z_t a)
{
  return a->sign && (a->chars[0] & 1);
}

inline static int zeven_nonzero(z_t a)
{
  return !(a->chars[0] & 1);
}

inline static int zodd_nonzero(z_t a)
{
  return a->chars[0] & 1;
}

inline static int zzero(z_t a)
{
  return !a->sign;
}

inline static int zsignum(z_t a)
{
  return a->sign;
}

extern z_t libzahl_tmp_cmp;
extern z_t libzahl_tmp_str_num;
extern z_t libzahl_tmp_str_mag;
extern z_t libzahl_tmp_str_div;
extern z_t libzahl_tmp_str_rem;
extern z_t libzahl_tmp_gcd_u;
extern z_t libzahl_tmp_gcd_v;
extern z_t libzahl_tmp_sub;
extern z_t libzahl_tmp_modmul;
extern z_t libzahl_tmp_div;
extern z_t libzahl_tmp_mod;
extern z_t libzahl_tmp_pow_b;
extern z_t libzahl_tmp_pow_c;
extern z_t libzahl_tmp_pow_d;
extern z_t libzahl_tmp_modsqr;
extern z_t libzahl_tmp_divmod_a;
extern z_t libzahl_tmp_divmod_b;
extern z_t libzahl_tmp_divmod_d;
extern z_t libzahl_tmp_ptest_x;
extern z_t libzahl_tmp_ptest_a;
extern z_t libzahl_tmp_ptest_d;
extern z_t libzahl_tmp_ptest_n1;
extern z_t libzahl_tmp_ptest_n4;
extern z_t libzahl_const_1e19;
extern z_t libzahl_const_1e9;
extern z_t libzahl_const_1;
extern z_t libzahl_const_2;
extern z_t libzahl_const_4;
extern z_t libzahl_tmp_divmod_ds[32];
extern jmp_buf libzahl_jmp_buf;
extern int libzahl_set_up;
extern int libzahl_error;
extern zahl_char_t **libzahl_pool[(sizeof(size_t)) * 8];
extern size_t libzahl_pool_n[(sizeof(size_t)) * 8];
extern size_t libzahl_pool_alloc[(sizeof(size_t)) * 8];
void libzahl_realloc(z_t a, size_t need);
enum zprimality zptest(z_t witness, z_t n, int t)
{
  size_t i;
  size_t r;
  if (zcmpu(n, 3) <= 0)
  {
    if (zcmpu(n, 1) <= 0)
    {
      if (witness)
      {
        do
        {
          if (witness != n)
          {
            zset(witness, n);
          }
        }
        while (0);
      }
      return NONPRIME;
    }
    else
    {
      return PRIME;
    }
  }
  if (zeven(n))
  {
    if (witness)
    {
      do
      {
        if (witness != n)
        {
          zset(witness, n);
        }
      }
      while (0);
    }
    return NONPRIME;
  }
  zsub_unsigned(libzahl_tmp_ptest_n1, n, libzahl_const_1);
  zsub_unsigned(libzahl_tmp_ptest_n4, n, libzahl_const_4);
  r = zlsb(libzahl_tmp_ptest_n1);
  zrsh(libzahl_tmp_ptest_d, libzahl_tmp_ptest_n1, r);
  while (t--)
  {
    helper_zptest_1(&i, witness, n, r);
  }

  return PROBABLY_PRIME;
}


typedef uint32_t zahl_char_t;
typedef struct 
{
  int sign;
  size_t used;
  size_t alloced;
  zahl_char_t *chars;
} z_t[1];
enum zprimality
{
  NONPRIME = 0,
  PROBABLY_PRIME,
  PRIME
};
enum zranddev
{
  FAST_RANDOM = 0,
  SECURE_RANDOM
};
enum zranddist
{
  QUASIUNIFORM = 0,
  UNIFORM
};
enum zerror
{
  ZERROR_ERRNO_SET = 0
};
void zsetup(jmp_buf);
void zunsetup(void);
void zfree(z_t);
void zswap(z_t, z_t);
size_t zsave(z_t, void *);
size_t zload(z_t, const void *);
void zset(z_t, z_t);
void zseti(z_t, long long int);
void zsetu(z_t, unsigned long long int);
int zcmp(z_t, z_t);
int zcmpi(z_t, long long int);
int zcmpu(z_t, unsigned long long int);
int zcmpmag(z_t, z_t);
void zadd(z_t, z_t, z_t);
void zsub(z_t, z_t, z_t);
void zmul(z_t, z_t, z_t);
void zmodmul(z_t, z_t, z_t, z_t);
void zdiv(z_t, z_t, z_t);
void zdivmod(z_t, z_t, z_t, z_t);
void zmod(z_t, z_t, z_t);
void zsqr(z_t, z_t);
void zmodsqr(z_t, z_t, z_t);
void zneg(z_t, z_t);
void zabs(z_t, z_t);
void zpow(z_t, z_t, z_t);
void zmodpow(z_t, z_t, z_t, z_t);
void zpowu(z_t, z_t, unsigned long long int);
void zmodpowu(z_t, z_t, unsigned long long int, z_t);
void zadd_unsigned(z_t, z_t, z_t);
void zsub_unsigned(z_t, z_t, z_t);
void zand(z_t, z_t, z_t);
void zor(z_t, z_t, z_t);
void zxor(z_t, z_t, z_t);
void znot(z_t, z_t);
void zlsh(z_t, z_t, size_t);
void zrsh(z_t, z_t, size_t);
void ztrunc(z_t, z_t, size_t);
int zbtest(z_t, size_t);
void zsplit(z_t, z_t, z_t, size_t);
size_t zbits(z_t);
size_t zlsb(z_t);
void zbset(z_t, z_t, size_t, int);
void zgcd(z_t, z_t, z_t);
enum zprimality zptest(z_t, z_t, int);
void zrand(z_t, enum zranddev, enum zranddist, z_t);
char *zstr(z_t, char *);
int zsets(z_t, const char *);
size_t zstr_length(z_t, unsigned long long int);
enum zerror zerror(const char **);
void zperror(const char *);
inline static void zinit(z_t a)
{
  a->alloced = 0;
  a->chars = 0;
}

inline static int zeven(z_t a)
{
  return (!a->sign) || (!(a->chars[0] & 1));
}

inline static int zodd(z_t a)
{
  return a->sign && (a->chars[0] & 1);
}

inline static int zeven_nonzero(z_t a)
{
  return !(a->chars[0] & 1);
}

inline static int zodd_nonzero(z_t a)
{
  return a->chars[0] & 1;
}

inline static int zzero(z_t a)
{
  return !a->sign;
}

inline static int zsignum(z_t a)
{
  return a->sign;
}

extern z_t libzahl_tmp_cmp;
extern z_t libzahl_tmp_str_num;
extern z_t libzahl_tmp_str_mag;
extern z_t libzahl_tmp_str_div;
extern z_t libzahl_tmp_str_rem;
extern z_t libzahl_tmp_gcd_u;
extern z_t libzahl_tmp_gcd_v;
extern z_t libzahl_tmp_sub;
extern z_t libzahl_tmp_modmul;
extern z_t libzahl_tmp_div;
extern z_t libzahl_tmp_mod;
extern z_t libzahl_tmp_pow_b;
extern z_t libzahl_tmp_pow_c;
extern z_t libzahl_tmp_pow_d;
extern z_t libzahl_tmp_modsqr;
extern z_t libzahl_tmp_divmod_a;
extern z_t libzahl_tmp_divmod_b;
extern z_t libzahl_tmp_divmod_d;
extern z_t libzahl_tmp_ptest_x;
extern z_t libzahl_tmp_ptest_a;
extern z_t libzahl_tmp_ptest_d;
extern z_t libzahl_tmp_ptest_n1;
extern z_t libzahl_tmp_ptest_n4;
extern z_t libzahl_const_1e19;
extern z_t libzahl_const_1e9;
extern z_t libzahl_const_1;
extern z_t libzahl_const_2;
extern z_t libzahl_const_4;
extern z_t libzahl_tmp_divmod_ds[32];
extern jmp_buf libzahl_jmp_buf;
extern int libzahl_set_up;
extern int libzahl_error;
extern zahl_char_t **libzahl_pool[(sizeof(size_t)) * 8];
extern size_t libzahl_pool_n[(sizeof(size_t)) * 8];
extern size_t libzahl_pool_alloc[(sizeof(size_t)) * 8];
void libzahl_realloc(z_t a, size_t need);
enum zprimality zptest(z_t witness, z_t n, int t)
{
  size_t i;
  size_t r;
  if (zcmpu(n, 3) <= 0)
  {
    if (zcmpu(n, 1) <= 0)
    {
      if (witness)
      {
        do
        {
          if (witness != n)
          {
            zset(witness, n);
          }
        }
        while (0);
      }
      return NONPRIME;
    }
    else
    {
      return PRIME;
    }
  }
  if (zeven(n))
  {
    if (witness)
    {
      do
      {
        if (witness != n)
        {
          zset(witness, n);
        }
      }
      while (0);
    }
    return NONPRIME;
  }
  zsub_unsigned(libzahl_tmp_ptest_n1, n, libzahl_const_1);
  zsub_unsigned(libzahl_tmp_ptest_n4, n, libzahl_const_4);
  r = zlsb(libzahl_tmp_ptest_n1);
  zrsh(libzahl_tmp_ptest_d, libzahl_tmp_ptest_n1, r);
  while (t--)
  {
    zrand(libzahl_tmp_ptest_a, FAST_RANDOM, UNIFORM, libzahl_tmp_ptest_n4);
    zadd_unsigned(libzahl_tmp_ptest_a, libzahl_tmp_ptest_a, libzahl_const_2);
    zmodpow(libzahl_tmp_ptest_x, libzahl_tmp_ptest_a, libzahl_tmp_ptest_d, n);
    if ((!zcmp(libzahl_tmp_ptest_x, libzahl_const_1)) || (!zcmp(libzahl_tmp_ptest_x, libzahl_tmp_ptest_n1)))
    {
      continue;
    }
    for (i = 1; i < r; i += 1)
    {
      zsqr(libzahl_tmp_ptest_x, libzahl_tmp_ptest_x);
      zmod(libzahl_tmp_ptest_x, libzahl_tmp_ptest_x, n);
      if (!zcmp(libzahl_tmp_ptest_x, libzahl_const_1))
      {
        if (witness)
        {
          zswap(witness, libzahl_tmp_ptest_a);
        }
        return NONPRIME;
      }
      if (!zcmp(libzahl_tmp_ptest_x, libzahl_tmp_ptest_n1))
      {
        break;
      }
    }

    if (i == r)
    {
      if (witness)
      {
        zswap(witness, libzahl_tmp_ptest_a);
      }
      return NONPRIME;
    }
  }

  return PROBABLY_PRIME;
}

enum zprimality helper_zptest_1(size_t * const i_ref, z_t witness, z_t n, size_t r)
{
  size_t i = *i_ref;
  zrand(libzahl_tmp_ptest_a, FAST_RANDOM, UNIFORM, libzahl_tmp_ptest_n4);
  zadd_unsigned(libzahl_tmp_ptest_a, libzahl_tmp_ptest_a, libzahl_const_2);
  zmodpow(libzahl_tmp_ptest_x, libzahl_tmp_ptest_a, libzahl_tmp_ptest_d, n);
  if ((!zcmp(libzahl_tmp_ptest_x, libzahl_const_1)) || (!zcmp(libzahl_tmp_ptest_x, libzahl_tmp_ptest_n1)))
  {
    continue;
  }
  for (i = 1; i < r; i += 1)
  {
    zsqr(libzahl_tmp_ptest_x, libzahl_tmp_ptest_x);
    zmod(libzahl_tmp_ptest_x, libzahl_tmp_ptest_x, n);
    if (!zcmp(libzahl_tmp_ptest_x, libzahl_const_1))
    {
      if (witness)
      {
        zswap(witness, libzahl_tmp_ptest_a);
      }
      return NONPRIME;
    }
    if (!zcmp(libzahl_tmp_ptest_x, libzahl_tmp_ptest_n1))
    {
      break;
    }
  }

  if (i == r)
  {
    if (witness)
    {
      zswap(witness, libzahl_tmp_ptest_a);
    }
    return NONPRIME;
  }
  *i_ref = i;
}

File: artifacts/libzahl-1.0/translate/src/zptest.rs
use crate::*;
use lazy_static::lazy_static;
use std::sync::Mutex;

pub enum Zprimality {
    PRIME,
    NONPRIME,
}

pub fn helper_zptest_1(
    i_ref: &mut usize,
    witness: Option<&mut Z>,
    n: &mut Z,
    r: usize,
) -> Zprimality {
    let mut i = *i_ref;
    
    {
        let mut a_guard = libzahl_tmp_ptest_a.lock().unwrap();
        let mut n4_guard = LIBZAHL_TMP_PTEST_N4.lock().unwrap();
        let const2_guard = LIBZAHL_CONST_2.lock().unwrap();
        
        zrand(&mut *a_guard, Zranddev::FastRandom, Zranddist::Uniform, &mut *n4_guard);
        let a_copy = Z {
            sign: a_guard.sign,
            used: a_guard.used,
            alloced: a_guard.alloced,
            chars: a_guard.chars.clone(),
        };
        zadd_unsigned(&mut *a_guard, &a_copy, &*const2_guard);
    }

    {
        let mut x_guard = LIBZAHL_TMP_PTEST_X.lock().unwrap();
        let mut a_guard = libzahl_tmp_ptest_a.lock().unwrap();
        let mut d_guard = LIBZAHL_TMP_PTEST_D.lock().unwrap();
        
        zmodpow(&mut *x_guard, &mut *a_guard, &mut *d_guard, n);
    }

    {
        let x_guard = LIBZAHL_TMP_PTEST_X.lock().unwrap();
        let const1_guard = libzahl_const_1.lock().unwrap();
        let n1_guard = libzahl_tmp_ptest_n1.lock().unwrap();
        
        if zcmp(&*x_guard, &*const1_guard) == 0 || zcmp(&*x_guard, &*n1_guard) == 0 {
            *i_ref = i;
            return Zprimality::PRIME;
        }
    }

    for _ in 1..r {
        {
            let mut x_guard = LIBZAHL_TMP_PTEST_X.lock().unwrap();
            let mut x_copy = Z {
                sign: x_guard.sign,
                used: x_guard.used,
                alloced: x_guard.alloced,
                chars: x_guard.chars.clone(),
            };
            zsqr(&mut *x_guard, &mut x_copy);
            let mut x_copy = Z {
                sign: x_guard.sign,
                used: x_guard.used,
                alloced: x_guard.alloced,
                chars: x_guard.chars.clone(),
            };
            zmod(&mut *x_guard, &mut x_copy, n);
        }

        {
            let x_guard = LIBZAHL_TMP_PTEST_X.lock().unwrap();
            let const1_guard = libzahl_const_1.lock().unwrap();
            
            if zcmp(&*x_guard, &*const1_guard) == 0 {
                if let Some(witness) = witness {
                    let mut a_guard = libzahl_tmp_ptest_a.lock().unwrap();
                    zswap(witness, &mut *a_guard);
                }
                *i_ref = i;
                return Zprimality::NONPRIME;
            }
        }

        {
            let x_guard = LIBZAHL_TMP_PTEST_X.lock().unwrap();
            let n1_guard = libzahl_tmp_ptest_n1.lock().unwrap();
            
            if zcmp(&*x_guard, &*n1_guard) == 0 {
                break;
            }
        }
        i += 1;
    }

    if i == r {
        if let Some(witness) = witness {
            let mut a_guard = libzahl_tmp_ptest_a.lock().unwrap();
            zswap(witness, &mut *a_guard);
        }
        *i_ref = i;
        return Zprimality::NONPRIME;
    }

    *i_ref = i;
    Zprimality::PRIME
}


pub fn zptest(witness: Option<&mut Z>, n: &mut Z, t: i32) -> Zprimality {
    let mut i = 0;
    let mut r;

    // Check if n <= 3
    if zcmpu(n, 3) <= 0 {
        if zcmpu(n, 1) <= 0 {
            if let Some(witness) = witness {
                if !std::ptr::eq(witness, n) {
                    zset(witness, n);
                }
            }
            return Zprimality::NONPRIME;
        } else {
            return Zprimality::PRIME;
        }
    }

    // Check if n is even
    if zeven(n) {
        if let Some(witness) = witness {
            if !std::ptr::eq(witness, n) {
                zset(witness, n);
            }
        }
        return Zprimality::NONPRIME;
    }

    // Perform the primality test
    {
        let mut n1_guard = libzahl_tmp_ptest_n1.lock().unwrap();
        let const1_guard = libzahl_const_1.lock().unwrap();
        zsub_unsigned(&mut *n1_guard, n, &*const1_guard);
    }

    {
        let mut n4_guard = LIBZAHL_TMP_PTEST_N4.lock().unwrap();
        let const4_guard = libzahl_const_4.lock().unwrap();
        zsub_unsigned(&mut *n4_guard, n, &*const4_guard);
    }

    {
        let n1_guard = libzahl_tmp_ptest_n1.lock().unwrap();
        r = zlsb(&*n1_guard);
    }

    {
        let mut d_guard = LIBZAHL_TMP_PTEST_D.lock().unwrap();
        let n1_guard = libzahl_tmp_ptest_n1.lock().unwrap();
        zrsh(&mut *d_guard, &*n1_guard, r);
    }

    let mut remaining_tests = t;
let mut witness_ref = witness;           // Option<&mut Z>

while remaining_tests > 0 {

    let result = helper_zptest_1(&mut i, witness_ref.take(), n, r);

    if matches!(result, Zprimality::NONPRIME) {
        return Zprimality::NONPRIME;
    }
    remaining_tests -= 1;
}

Zprimality::PRIME
}
